File: app/src/main/java/com/example/purepawapp/ui/admin/AdminProductAdapter.java
package com.example.purepawapp.ui.admin;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.example.purepawapp.R;
import com.example.purepawapp.data.model.Product;
import com.example.purepawapp.databinding.ItemAdminProductBinding;
import com.example.purepawapp.ui.product.ProductUi;
import com.example.purepawapp.util.CurrencyUtils;

import java.util.function.Consumer;

public class AdminProductAdapter extends ListAdapter<Product, AdminProductAdapter.ViewHolder> {

    private final Consumer<Product> onDelete;

    private static final DiffUtil.ItemCallback<Product> DIFF_CALLBACK = new DiffUtil.ItemCallback<>() {
        @Override
        public boolean areItemsTheSame(@NonNull Product oldItem, @NonNull Product newItem) {
            return oldItem.getId().equals(newItem.getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull Product oldItem, @NonNull Product newItem) {
            return oldItem.equals(newItem);
        }
    };

    public AdminProductAdapter(Consumer<Product> onDelete) {
        super(DIFF_CALLBACK);
        this.onDelete = onDelete;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemAdminProductBinding binding = ItemAdminProductBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private final ItemAdminProductBinding binding;

        ViewHolder(ItemAdminProductBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(Product product) {
            binding.tvIcon.setText(ProductUi.getEmoji(product));
            binding.tvName.setText(product.getName());
            binding.tvCategory.setText(AdminStatusUi.categoryDisplayName(product.getCategoryId()));
            binding.tvPrice.setText(CurrencyUtils.toVndString(product.getDisplayPrice()));

            int totalStock = 0;
            for (var variant : product.getVariants()) totalStock += variant.getStock();
            boolean inStock = totalStock > 0;
            binding.tvStockStatus.setText(inStock ? "Còn hàng" : "Hết hàng");
            binding.tvStockStatus.setTextColor(binding.getRoot().getResources().getColor(
                    inStock ? R.color.pp_success_dark : R.color.pp_status_red_text, null));
            binding.tvStockStatus.setBackgroundTintList(binding.getRoot().getResources().getColorStateList(
                    inStock ? R.color.pp_chip_green_bg : R.color.pp_status_red_bg, null));

            binding.btnDelete.setOnClickListener(v -> onDelete.accept(product));
        }
    }
}

File: app/src/main/java/com/example/purepawapp/ui/admin/AdminProductAdapter.kt
package com.example.purepawapp.ui.admin

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.purepawapp.R
import com.example.purepawapp.data.model.Product
import com.example.purepawapp.databinding.ItemAdminProductBinding
import com.example.purepawapp.ui.product.emoji
import com.example.purepawapp.util.toVndString

class AdminProductAdapter(
    private val onDelete: (Product) -> Unit
) : ListAdapter<Product, AdminProductAdapter.ViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = ItemAdminProductBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class ViewHolder(private val binding: ItemAdminProductBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(product: Product) {
            binding.tvIcon.text = product.emoji
            binding.tvName.text = product.name
            binding.tvCategory.text = categoryDisplayName(product.categoryId)
            binding.tvPrice.text = product.displayPrice.toVndString()

            val totalStock = product.variants.sumOf { it.stock }
            val inStock = totalStock > 0
            binding.tvStockStatus.text = if (inStock) "Còn hàng" else "Hết hàng"
            binding.tvStockStatus.setTextColor(
                binding.root.resources.getColor(
                    if (inStock) R.color.pp_success_dark else R.color.pp_status_red_text,
                    null
                )
            )
            binding.tvStockStatus.backgroundTintList = binding.root.resources.getColorStateList(
                if (inStock) R.color.pp_chip_green_bg else R.color.pp_status_red_bg,
                null
            )

            binding.btnDelete.setOnClickListener { onDelete(product) }
        }
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<Product>() {
            override fun areItemsTheSame(oldItem: Product, newItem: Product): Boolean = oldItem.id == newItem.id
            override fun areContentsTheSame(oldItem: Product, newItem: Product): Boolean = oldItem == newItem
        }
    }
}
